src/Files.App/Services/Windows/WindowsRecentItemsService.cs

```diff
@@ -0,0 +1,318 @@
+	public class WindowsRecentItemsService : IWindowsRecentItemsService
+	{
+		// Dependency injections
+
+		private readonly IFoldersSettingsService FoldersSettingsService = Ioc.Default.GetRequiredService<IFoldersSettingsService>();
+
+		// Fields
+
+		private readonly SystemIO.FileSystemWatcher? _watcher;
+
+		// Properties
+
+		private readonly List<RecentItem> _RecentFiles = [];
+		/// <inheritdoc/>
+		public IReadOnlyList<RecentItem> RecentFiles
+		{
+			get
+			{
+				lock (_RecentFiles)
+					return _RecentFiles.ToList().AsReadOnly();
+			}
+		}
+
+		private readonly List<RecentItem> _RecentFolders = [];
+		/// <inheritdoc/>
+		public IReadOnlyList<RecentItem> RecentFolders
+		{
+			get
+			{
+				lock (_RecentFolders)
+					return _RecentFolders.ToList().AsReadOnly();
+			}
+		}
+
+		// Events 
+
+		/// <inheritdoc/>
+		public event EventHandler<NotifyCollectionChangedEventArgs>? RecentFilesChanged;
+
+		/// <inheritdoc/>
+		public event EventHandler<NotifyCollectionChangedEventArgs>? RecentFoldersChanged;
+
+		// Constructor
+
+		public WindowsRecentItemsService()
+		{
+			_watcher = new()
+			{
+				Path = SystemIO.Path.Combine(Environment.GetFolderPath(Environment.SpecialFolder.Recent), "AutomaticDestinations"),
+				Filter = "5f7b5f1e01b83767.automaticDestinations-ms",
+				NotifyFilter = SystemIO.NotifyFilters.DirectoryName | SystemIO.NotifyFilters.FileName | SystemIO.NotifyFilters.LastWrite,
+			};
+
+			_watcher.Changed += Watcher_Changed;
+			_watcher.Deleted += Watcher_Changed;
+			_watcher.EnableRaisingEvents = true;
+		}
+
+		// Methods
+
+		/// <inheritdoc/>
+		public async Task<bool> UpdateRecentFilesAsync()
+		{
+			return await Task.Run(() =>
+			{
+				return UpdateRecentItems(false);
+			});
+		}
+
+		/// <inheritdoc/>
+		public async Task<bool> UpdateRecentFoldersAsync()
+		{
+			return await Task.Run(() =>
+			{
+				return UpdateRecentItems(true);
+			});
+		}
+
+		/// <inheritdoc/>
+		public unsafe bool Add(string path)
+		{
+			try
+			{
+				fixed (char* cPath = path)
+					PInvoke.SHAddToRecentDocs((uint)SHARD.SHARD_PATHW, cPath);
+
+				return true;
+			}
+			catch (Exception ex)
+			{
+				App.Logger.LogWarning(ex, ex.Message);
+				return false;
+			}
+		}
+
+		/// <inheritdoc/>
+		public unsafe bool Remove(RecentItem item)
+		{
+			try
+			{
+				var bhid = PInvoke.BHID_SFUIObject;
+				var contextMenuIid = typeof(IContextMenu).GUID;
+				using ComPtr<IContextMenu> pContextMenu = default;
+				HRESULT hr = item.ShellItem.Get()->BindToHandler(null, &bhid, &contextMenuIid, (void**)pContextMenu.GetAddressOf());
+				HMENU hMenu = PInvoke.CreatePopupMenu();
+				hr = pContextMenu.Get()->QueryContextMenu(hMenu, 0, 1, 0x7FFF, PInvoke.CMF_OPTIMIZEFORINVOKE);
+
+				// Initialize invocation info
+				CMINVOKECOMMANDINFO cmi = default;
+				cmi.cbSize = (uint)sizeof(CMINVOKECOMMANDINFO);
+				cmi.nShow = (int)SHOW_WINDOW_CMD.SW_HIDE;
+
+				// Try unpin first for pinned files
```

Comment:
I'm assuming this includes folders too? If so we should execute both `unpinfromhome` and `removefromhome`. The former doesn't work for a lot of the default Windows folders like Music & Video.

https://github.com/files-community/Files/commit/ba6624b22e818832d1cb3fbfe9f20bdca91d0504#diff-e7c944353f705cd45acb6c334f0ce1693abda765b0217b207ef03c71c96e2143R180

Reply:
Oh I was confused by these. These verbs are undocumented?
Is there an unofficial documentation about this?

I have an idea to avoid verbs but as I mentioned above we don't plan for. Ow fyi.

Reply:
I was never able to find official docs for this outside of forums, but in my own testing it's feasible & safe to call both if you want to remove a **folder.** There's *probably* not a need to invoke these verbs for files though.

Reply:
I see. I'll resolve it shortly.

Reply:
@jiejasonliu I just should execute both unpinfromhome and removefromhome? Not necessarily `remove`?

Reply:
So `remove` is still needed to remove files (non-folders).
If we're trying to unpin **folders**, then I suggest executing both `unpinfromhome` and `removefromhome`.

This flow does not know whether the recent item we're handling is a file/folder right?
I'm not sure what happens if you invoke the wrong verbs for the wrong type (file vs. folder), but my guess is nothing and maybe we could fire all three if it's too difficult to distinguish what type of recent item we're handling.

Reply:
> This flow does not know whether the recent item we're handling is a file/folder right?

Yeah, should I add IsFolder to WidgetRecentItem at least?

> I'm not sure what happens if you invoke the wrong verbs for the wrong type (file vs. folder), but my guess is nothing and maybe we could fire all three if it's too difficult to distinguish what type of recent item we're handling.

I guess the invocation method would fail with E_FAIL or some HRESULT.

> So remove is still needed to remove files (non-folders).

Got it. I'll add invocation code of these 3 verbs